External/FEXCore/include/FEXCore/Utils/refcount_shared_mutex.h

```diff
@@ -0,0 +1,327 @@
+            }
+          }
+
+          return std::make_pair(LocalFutex, false);
+        };
+
+        auto UniqueResult = TryUniqueLock();
+        return UniqueResult.second;
+      }
+
+      void unlock() {
+        LOGMAN_THROW_A_FMT(Futex.load() == UNIQUE_LOCK_VALUE, "Tried unlocking not locked mutex?");
+
+        auto TryUniqueUnlock = [this]() -> std::pair<uint32_t, bool> {
+          auto LocalFutex = Futex.load();
+
+          if (LocalFutex != UNIQUE_LOCK_VALUE) {
+            // Refcount must be zero if we are to attempt getting a unique lock
+          }
+          else {
+            // Try locking now in userspace
+            while (LocalFutex == UNIQUE_LOCK_VALUE) {
+              auto Desired = LocalFutex;
+              Desired = 0;
+              if (Futex.compare_exchange_strong(LocalFutex, Desired)) {
+                // We have successfully unique locked
+                return std::make_pair(Desired, true);
+              }
+              else {
+                if (LocalFutex == UNIQUE_LOCK_VALUE) {
+                  // If another thread pulled the unique lock or the ref count incremented
+                  // Then we need to wait, loop will end
+                }
+              }
+            }
+          }
+
+          return std::make_pair(LocalFutex, false);
+        };
+
+        [[maybe_unused]] auto UniqueResult = TryUniqueUnlock();
+        LOGMAN_THROW_A_FMT(UniqueResult.second, "Couldn't unlock mutex memory?");
+
+        // We've now unlocked, use the futex to wake up any shared waiters
+        int Op = FUTEX_WAKE | FUTEX_PRIVATE_FLAG;
+        ::syscall(SYS_futex,
+          &Futex,
+          Op,
+          INT_MAX, // Could be any number of shared waiters
+          nullptr, // timeout
+          nullptr, // addr
+          0);
+      }
+
+      bool try_lock_shared() {
+        auto TryRefIncrement = [this]() -> bool {
+          auto LocalFutex = Futex.load();
+
+          if (LocalFutex == UNIQUE_LOCK_VALUE) {
+            // Unique lock held
+          }
+          else {
+            // Try to increment the counter if unique lock isn't held
+            while (LocalFutex != UNIQUE_LOCK_VALUE) {
+              auto Desired = LocalFutex;
+              Desired++;
+
+              // Try to increment the ref count
+              if (Futex.compare_exchange_strong(LocalFutex, Desired)) {
+                // We have successfully incremented the ref counting mutex in userspace
+                return true;
+              }
+              else {
+                if (LocalFutex == UNIQUE_LOCK_VALUE) {
+                  // Unique lock was held
+                  // Nothing to do, loop will end
+                }
+
+                // Try again. Can happen in a race to increment the ref count
+              }
+            }
+          }
+
+          return false;
+        };
+
+        return TryRefIncrement();
+      }
+
+      void lock_shared() {
+        auto TryRefIncrement = [this]() -> bool {
```

Comment:
This seems to be duplicated across many of the member functions. Why not move it to a private member function (generalizing it a bit if need be?)

Reply:
Changed over these lambdas to private member functions. Resolved.